Bump djoser to 2.3.1

marslanabdulrauf opened the pull request:
### What are the relevant tickets?
https://github.com/mitodl/mitxpro/security/dependabot/204

### Description (What does it do?)
This PR upgrades djoser to 2.3.1

Anas12091101 reviewed (approved):
LGTM 👍 Tested in xPRO

marslanabdulrauf commented:
**Test Steps**
**Direct Approach**:
1. Updated the `djoser` dependency in `xpro` to match the new version.
3. Ran the application locally to ensure there were no breaking changes.
4. Verified the authentication flow:
    - Checked user login/logout functionality.
    - Validated user registration and password reset flows.
    - Confirmed email token generation and validation.

Reviewed the test suite and ensured all tests passed successfully.

**Testing with Local Changes in `mitol-django-authentication`:**
1. Clone the `mitol-django-authentication` repository locally.
2. Checked out the branch where `djoser` was updated to >=2.3.0.
3. In the xpro project:
    - Removed the existing `mitol-django-authentication` package.
    - Installed the local version of the updated `mitol-django-authentication` package in the Docker container.
5. Ran the application locally and verified:
    - Authentication flows (login/logout, registration, password reset).
    - Compatibility of djoser functionality with existing xPro code.

Run the xpro test suite and ensured all tests passed.

Anas12091101 commented:
I tested it by directly updating the package to v2.3.1 in xPRO

rhysyngsun commented:
`python-saml` I think is only necessary for touchstone integrations, so if that package has code around that that's probably why it's used.

rhysyngsun commented:
@marslanabdulrauf go for it

marslanabdulrauf commented:
Thank you @arslanashraf7 for highlighting the OCW Studio, I completely missed that. I will check with OCW Team for confirmation before merging this.

marslanabdulrauf commented:
I had a discussion over a huddle with @umar8hassan from the OCW team, and here are our findings:

- **OCW-Studio** is using `mitol-django-authentication` version `2023-12-19`, which already excludes `python3-saml`, but it still uses **Touchstone** for login.
- In OCW-Studio, `python3-saml` is actually being pulled in from `social-auth-core[saml]` as seen in:
    - [pyproject.toml reference](https://github.com/mitodl/ocw-studio/blob/master/pyproject.toml#L51)
    - [poetry.lock reference](https://github.com/mitodl/ocw-studio/blob/master/poetry.lock#L3198)

Given this, removing `python3-saml` from `mitol-django-authentication` should not affect **OCW-Studio**, as it's still receiving `python3-saml` via `social-auth-core[saml]`

However, since this PR also removes **Touchstone-related** settings and the `saml_metadata` view, we need confirmation from the OCW team on whether those settings are required for **OCW-Studio.**

I’m attaching the latest package build for review: [mitol_django_authentication-2025.1.9.tar.gz](https://github.com/user-attachments/files/18620559/mitol_django_authentication-2025.1.9.tar.gz).

Can someone from OCW verify this?

Cc @pdpinch

pdpinch commented:
> However, since this PR also removes Touchstone-related settings and the saml_metadata view, we need confirmation from the OCW team on whether those settings are required for OCW-Studio.

@umar8hassan can you take a look at this?

umar8hassan commented:
> > However, since this PR also removes Touchstone-related settings and the saml_metadata view, we need confirmation from the OCW team on whether those settings are required for OCW-Studio.
> 
> @umar8hassan, can you take a look at this?

@marslanabdulrauf @pdpinch We do use `touchstone` directly in OCW settings module.

![image](https://github.com/user-attachments/assets/647a48db-be46-4d97-b35b-c8d2ad3e54a0)

No easy way to disintegrate touchstone.

<img width="960" alt="Screenshot 2025-02-06 at 6 45 11 PM" src="https://github.com/user-attachments/assets/11792b63-d5ac-4683-91f9-7d273fe52f8f" />

pdpinch commented:
Ok. Then we shouldn't remove those files? 

It's not clear to me why these would be removed in a PR that is' updating djoser anyway.

marslanabdulrauf commented:
> Ok. Then we shouldn't remove those files?

Yes, we will not remove these files for now. I will update the xPro application to resolve the error mentioned in this [comment](https://github.com/mitodl/ol-django/pull/175#pullrequestreview-2566445967).

> It's not clear to me why these would be removed in a PR that is' updating djoser anyway.

Regarding the removal of `python3-saml`, I have provided an explanation in this comment: https://github.com/mitodl/ol-django/pull/175#issuecomment-2615819016

marslanabdulrauf commented:
Thanks @arslanashraf7 for your extensive testing

> We need to make similar changes (https://github.com/mitodl/mitxpro/pull/3392) to other applications that are using the authentication because if they don't have no-binary packages they will break when the new build for the authentication package gets on PyPI.

I have performed a [basic search](https://github.com/search?q=+mitol-django-authentication&type=code) and found out that two of our repos are using this package:
1. [mitxonline](https://github.com/mitodl/mitxonline/blob/main/poetry.lock#L2135)
2. [bootcamp-ecommerce](https://github.com/mitodl/bootcamp-ecommerce/blob/master/poetry.lock#L2039)

**Safe:**
The have pinned the version for this `mitol-django-authentication` package

**Next steps?**
I have seen that both of these repositories also have security issue related to `djoser` 
1. [mitxonline security (high)](https://github.com/mitodl/mitxonline/security/dependabot/246)
2. [bootcamp-ecommerce security (high)](https://github.com/mitodl/bootcamp-ecommerce/security/dependabot/217)

We can follow them as required/necessary
